### DISCUSSION TITLE
Allowing both SSL and non SSL connections

### DISCUSSION THREAD

**Discussion by kiropowered** [quickfix-j/quickfixj#760] [Q&A]
Hi,
I would like to handle non ssl connection to a specific port and ssl on another one, i also want both of them to handle dynamic sessions.
So I:
 - I used 2 different implementations of AcceptorSessionProvider one for ssl and one for no ssl.
 - Binded them to 2 different ports with method acceptor.setSessionProvider
 - In the ssl provider I make sure to define the setting truststore
 - I added 2 templates session as follow : 
 
``` 
[session]
AcceptorTemplate=Y
BeginString=FIX.4.4
SocketAcceptPort=9877

[session]
AcceptorTemplate=Y
SocketUseSSL=Y
CipherSuites=TLS_RSA_WITH_AES_128_CBC_SHA
EnabledProtocols=TLSv1.2
SocketKeyStore=acceptor.keystore
SocketKeyStorePassword=password
NeedClientAuth=Y
BeginString=FIX.4.4
SocketAcceptPort=9878
```

It seems like the last definition override the first one and the app only listen to 9878 even thoI binded the 2 ports in  acceptor.setSessionProvider.

Is it possible to have both ssl and non ssl ?

**Comment by chrjohn** [comment]
Hmm, did not really look into it, but maybe you might want to assign unique ports? In your config you assign 9877 twice...

**Reply by kiropowered** [reply]
Sry I just did an error when editing the ticket when I removed "sensitive" info but they are actually on different ports. I edited my original post. I think you can reproduce the issue with the config I mentioned and see that only the last definition is used.

Feels like when there is 2 times the same BeginString on a template only the last one is actually used

**Reply by the-thing** [reply]
@kiropowered Could you show the full configuration including the common section? You don't need to provide values for fields that expose some internal information, just key names e.g Key=

**Comment by kiropowered** [comment]
Sure this is the config, 

`[default]
FileStorePath
ConnectionType=acceptor
SenderCompID=*
TargetCompID=*
StartTime=00:00:00
EndTime=00:00:00
FileLogPath=
RefreshOnLogon=Y
TestRequestDelayMultiplier=1
UseDataDictionary=Y
DataDictionary=dictionnary


JdbcDriver=com.mysql.cj.jdbc.Driver
JdbcURL=jdbc:mysql:
JdbcUser=
JdbcPassword=
JdbcLogHeartBeats=N
JdbcStoreMessagesTableName=
JdbcStoreSessionsTableName=
JdbcLogEventTable=
JdbcLogIncomingTable=
JdbcLogOutgoingTable=

[session]
AcceptorTemplate=Y
BeginString=FIX.4.4
SocketAcceptPort=9877

[session]
AcceptorTemplate=Y
SocketUseSSL=Y
CipherSuites=
EnabledProtocols=
SocketKeyStore=
SocketKeyStorePassword=
NeedClientAuth=Y
BeginString=FIX.4.4
SocketAcceptPort=9878
`

I also call of course 2 times with the corresponding ports 9877 and 9878, providing my custom acceptors
acceptor.setSessionProvider

But as I said it only listen to the last definition wich is 9878, if i switch position then it's the port 9877 that is used, but i never can get both to work

**Reply by the-thing** [reply]
@kiropowered 

I never tried this scenario so I wrote a test case https://github.com/quickfix-j/quickfixj/pull/761 that demonstrates that having multiple sessions with ssl enabled/disabled is in fact working. It seems that it is probably something with your application setup. Configuration you provided looks ok to me, but it seems you have some custom parameters that require custom code.

If you think this is a problem with QuickFIX and not your application maybe you could write a test case quickfixj-core/src/test/java/quickfix/mina/ssl/SSLCertificateTest.java

BTW. When NeedClientAuth=Y you also need to provide a truststore in acceptor otherwise all initiatiors will be trusted.

@chrjohn 

It might be worth merging this test case https://github.com/quickfix-j/quickfixj/pull/761/

**Reply by kiropowered** [reply]
Your tests use static sessions and i am using dynamic.
In my case I use 2 sessionProvider one for the ssland one for the non ssl. with for both sessions
AcceptorTemplate=Y
TargetCompID=* 
SenderCompID=*
BeginString=FIX.4.4

I think it leads to the issue I mentionned cause maybe I am wrong but the tests does not cover my case

**Reply by the-thing** [reply]
Are you saying that the issue is with your implementation of custom providers? We don't see the code so it is going to be hard to advise what is wrong. If you can code a failing test case this would be helpful to see if actually something is wrong with quickfix.

**Reply by kiropowered** [reply]
I copied part of your test code and edited it a bit to give you an idea, I removed some DI to keep it clean.
But the idea here is to have dynamic session in the non ssl and ssl socket

Not sure how to represent my config in your createMixedSessionAcceptorSettings method, cause it use template session and there is no Target and Sender wich are definied as * in default

**Reply by the-thing** [reply]
I don't think I will be able to help if you don't provide minimal working example.

**Reply by kiropowered** [reply]
Sorry I don't have a lot of experience in Java, I can't make the tests running I tried with 

`./mvnw test`

It leads to the following error: 

`Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.6.1:unpack (unpack-dependencies) on project quickfixj-messages-fixt11: Artifact has not been packaged yet. When used on reactor artifact, unpack should be executed after packaging: see MDEP-98. -> [Help 1]`

Doing 

`mvnw clean package -Dmaven.javadoc.skip=true -PskipBundlePlugin,minimal-fix-latest`

It takes an eternity and never end.

**Reply by the-thing** [reply]
Try building the whole project in your IDE. I build mine with "mvn clean package -Dmaven.javadoc.skip=true -DskipTests -PskipBundlePlugin" command in IDEA. If the build is successful you will be able to run a single test case from your IDE. Once it works write a minimal working example demonstrating what you are trying to do and why it should work.

If you have problems with the above, you have to solve them one by one by yourself. Also it is not good idea to paste the code as probably nobody is going to look at those. It might be too long and complex.

**Reply by chrjohn** [reply]
Thank you @the-thing . Just wanted to write something similar. 
@kiropowered  you need to make it easy for other people to help you. I think the least you could do is to create a branch with your changes and a minimal reproducer.
Thank you

**Reply by kiropowered** [reply]
Ok this is my commit : 
https://github.com/kiropowered/quickfixj/commit/6dc355e250cd011d78a963331b3b211f81be819d

Added 2 tests methods:
shouldConnectDifferentTypesOfDynamicSessionsFail
shouldConnectDifferentTypesOfDynamicSessionsSuccess


Seems like if I set a session qualifier for the acceptor it actually works and create the 2 sockets(ssl/nossl) otherwise the config is merged and it listens to only the last specified port

**Reply by the-thing** [reply]
I had a brief look at the code. Yes. With the failing scenario the session settings you are creating are replacing each other.

https://github.com/kiropowered/quickfixj/commit/6dc355e250cd011d78a963331b3b211f81be819d#diff-2e6ec7295da7919787ed7d3829b8c86e72c135dad8689470dd130173f9073902R1009

https://github.com/kiropowered/quickfixj/commit/6dc355e250cd011d78a963331b3b211f81be819d#diff-2e6ec7295da7919787ed7d3829b8c86e72c135dad8689470dd130173f9073902R1000

Which causes the ssl socket and non ssl socket mixed up. Port is not something that differentiates the sessions. 

Seems like you found the issue. Is this something that solves your problem?

**Reply by kiropowered** [reply]
What do you mean by "is this something that solves your problem ?".

The SessionQualifier setting is mentionned in documentation as following : 
"This can only be used with initiator sessions"

Didn't thought of adding it in the acceptor config tbh.

**Reply by the-thing** [reply]
You problem statement was that you can't use ssl and not ssl dynamic sessions. Now it seems that you are able to.

Not sure if the documentation is up to date.

**Reply by chrjohn** [reply]
You should rather not use the SessionQualifier.

Please take this with a grain of salt because it's been a while since I last used dynamic sessions:
The way I see it, QFJ treats both sessions the same because they have the same CompIDs. Actually it would be sufficient to use `AcceptorTemplate=Y` for both session definitions and then give both session definitions different CompIDs, e.g. A and B.
Then just pass the `DynamicAcceptorSessionProvider` (you said you have two) different `templateIDs`.
If that does not work maybe you need to implement a custom provider.

**Reply by kiropowered** [reply]
The whole idea behind using dynamic sessions is to not have to edit config files and restart the app to add/edit sessions.
Having both ssl/nossl mode available for our clients while let them choose if they want a secure connection or not(using the corresponding port).
Of course the CompID will be different from ssl to non ssl with a suffix like _SSL.

But as you can see even with different compids the test fail with AcceptorTemplate=Y on both session definition cause one override the other based on 
BeginString=FIX.4.4 
SenderCompID=*
TargetCompID=*



The only solution seems like to add SessionQualifier in one ofthem in order for the parser to actually listen to both ports.

So I am not sure to follow your solution

**Reply by chrjohn** [reply]
> But as you can see even with different compids the test fail with AcceptorTemplate=Y on both session definition cause one override the other based on BeginString=FIX.4.4 SenderCompID=* TargetCompID=*

Using * on both sessions as CompIDs are no different CompIDs.
I meant that your config should look like this:

```
[default]
...
ConnectionType=acceptor
BeginString=FIX.4.4
...

[session]
AcceptorTemplate=Y
SenderCompID=A
TargetCompID=A
SocketAcceptPort=9877

[session]
AcceptorTemplate=Y
SenderCompID=B
TargetCompID=B
SocketAcceptPort=9878
... ssl settings and so on ...
```

Then do NOT use this constructor which defines one single template mapping with *

https://github.com/quickfix-j/quickfixj/blob/8e25819914498cee8568a68d9909f02f59b2aab2/quickfixj-core/src/main/java/quickfix/mina/acceptor/DynamicAcceptorSessionProvider.java#L97-L102

Use this constructor which allows to pass different template mappings, in your case A and B:

https://github.com/quickfix-j/quickfixj/blob/8e25819914498cee8568a68d9909f02f59b2aab2/quickfixj-core/src/main/java/quickfix/mina/acceptor/DynamicAcceptorSessionProvider.java#L120-L123

**Reply by kiropowered** [reply]
I actually don't use the DynamicAcceptorSessionProvider, I use a custom made one that look for sessions config in db.
That's what allow us to be fully dynamic and do some authorization so not any session is authorized and load keys if session are using ssl etc...

In your config file you set session with hardcoded values in SenderCompID and TargetCompID I still don't understand why

**Reply by chrjohn** [reply]
Then please show the code of your provider. Maybe in your branch.

Look at the code of DynamicAcceptorSessionProvider. It looks for the specified wildcard from the settings (be it A, B or *) and copies the settings from that template session to the newly created sessions. I suspect that you need something like that as well.
All that `CompID=*` stuff in the settings is only actually used by the DynamicAcceptorSessionProvider. If you don't use that provider it is pretty pointless to have * as CompIDs.

**Reply by kiropowered** [reply]
Ok so you advocate to use 2 differents compIds instead of using wildcard in both session with acceptorQualifier to have 2 listening socket. 
Using wildcards look more appropriate in term of "configuration visualization" like it means "anything" usually.
